includes/data-events/class-data-events.php

```diff
@@ -0,0 +1,230 @@
+<?php
+/**
+ * Newspack Data Events.
+ *
+ * @package Newspack
+ */
+
+namespace Newspack;
+
+use WP_Error;
+
+/**
+ * Main class.
+ */
+final class Data_Events {
+	/**
+	 * Asynchronous action name.
+	 */
+	const ACTION = 'newspack_data_event';
+
+	/**
+	 * Registered callable handlers, keyed by their action name.
+	 *
+	 * @var callable[]
+	 */
+	private static $actions = [];
+
+	/**
+	 * Registered global callable handlers to be executed on all actions.
+	 *
+	 * @var callable[]
+	 */
+	private static $global_handlers = [];
+
+	/**
+	 * Initialize hooks.
+	 */
+	public static function init() {
+		\add_action( 'wp_ajax_' . self::ACTION, [ __CLASS__, 'maybe_handle' ] );
+		\add_action( 'wp_ajax_nopriv_' . self::ACTION, [ __CLASS__, 'maybe_handle' ] );
+	}
+
+	/**
+	 * Maybe handle an event.
+	 */
+	public static function maybe_handle() {
+		// Don't lock up other requests while processing.
+		session_write_close(); // phpcs:ignore
+
+		if ( ! isset( $_REQUEST['nonce'] ) || ! \wp_verify_nonce( \sanitize_text_field( $_REQUEST['nonce'] ), self::ACTION ) ) {
```

Comment:
You can use `check_ajax_referrer` here. a bit simpler -> https://developer.wordpress.org/reference/functions/check_ajax_referer/

Reply:
I opted for a more verbose approach for readability. That function feels like an unnecessary shortcut.